widgets/src/main/java/org/pentaho/gwt/widgets/client/listbox/CustomListBox.java

```diff
@@ -1027,8 +1030,11 @@ public void onKeyUp( Widget widget, char c, int i ) {
         break;
       case 27: // ESC
       case 13: // Enter
-        if ( popupShowing ) {
-          togglePopup();
+      case 32: // Space
+        if ( !popupShowing ) {
+          this.togglePopup();
+        } else {
+          this.popup.hide();
```

Comment:
Just calling `this.togglePopup();` would do, for both when the popup is showing and not.
This method already checks for `popupShowing` and hides the popup when showing.

Reply:
Also, one other issue is, why would pressing ESC open the popup?
You should probably split the handling of these keys.

```java
  case 27: // ESC
     if ( popupShowing ) {
       togglePopup();
    }
    break;
  
  case 13: // Enter
  case 32: // Space
    this.togglePopup();
    break;
```

Reply:
> One other issue is, why would pressing ESC open the popup? You should probably split the handling of these keys.
> 
> ```java
>   case 27: // ESC
>     if ( popupShowing ) {
>       togglePopup();
>     }
>     break;
>   
>   case 13: // Enter
>   case 32: // Space
>     this.togglePopup();
>     break;
> ```

This line was already present in the code, anyway I've tested after removing the Esc case makes no difference in the behaviour (closing the dialog).

Reply:
Before, ESC was only taking effect when hiding the popup. Your proposed code changed started opening the popup when hidden.

Reply:
And the newly proposed code completely removed the ESC key handling. AFAIK, the ESC key should close the dialog.

Above, I proposed:
```java
case 27: // ESC
    if ( popupShowing ) {
      this.togglePopup();
    }
    break;
```

But you did not include that.

Reply:
There was an old bug which I was trying to troubleshoot for which I had to create an override method. Will be adding that code change back like you suggested.